### PR TITLE
[FLINK-20877][table-runtime-blink] Refactor BytesHashMap and BytesMultiMap to support window key

### PR DESCRIPTION
**Pull request by wuchong**
## What is the purpose of the change

Currently, the BytesHashMap and BytesMultiMap only support BinaryRowData as the key. However, for window operators, the key should be groupKey + window. We should make them decouple with BinaryRowData key type and support window key.

## Brief change log

0. Move `BytesMap` classes to util packag. Because in the near future, the structures will not only be used in aggregations, but also topn and others. 
2. Extract pages operations from `AbstractRowDataSerializer` into `PagedTypeSerializer`
1. Make `BytesMap` decouple with `BinaryRowData` key type
3. Introduce `WindowKey` structure and `WindowKeySerializer`
4. Introduce `BytesWindowHashMap` and `BytesWindowMultiMap` to support window key
We still keep the original name of `BytesHashMap` unchanged to avoid changing too many codes.

## Verifying this change

Refactor `BytesHashMapTest` and `BytesMultiMapTest` tests to also cover the new introduced `BytesWindowHashMap` and `BytesWindowMultiMap`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (**yes** / no / don't know)
  - The runtime per-record code paths (performance sensitive): (**yes** / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/typeutils/BinaryRowDataSerializer.java [hunk 1]
```diff
@@ -197,6 +197,11 @@ public BinaryRowData mapFromPages(BinaryRowData reuse, AbstractPagedInputView he
         return reuse;
     }
 
+    @Override
+    public void skipRecordToRead(AbstractPagedInputView source) throws IOException {
+        source.skipBytes(source.readInt());
```
Comment (JingsongLi) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/typeutils/BinaryRowDataSerializer.java:202]: Why not `checkSkipReadForFixLengthPart`?

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/typeutils/BinaryRowDataSerializer.java [hunk 2]
```diff
@@ -197,6 +197,11 @@ public BinaryRowData mapFromPages(BinaryRowData reuse, AbstractPagedInputView he
         return reuse;
     }
 
+    @Override
+    public void skipRecordToRead(AbstractPagedInputView source) throws IOException {
```
Comment (JingsongLi) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/typeutils/BinaryRowDataSerializer.java:201]: A better name to remind `pages`?

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/typeutils/PagedTypeSerializer.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+     * <p>If you choose the zero copy way, you have to deal with the lifecycle of the pages
+     * properly.
+     *
+     * @param source The input view from which to read the data.
+     * @return The mapped element.
+     * @throws IOException Thrown, if the de-serialization encountered an I/O related error.
+     *     Typically raised by the input view, which may have an underlying I/O channel from which
+     *     it reads.
+     */
+    public abstract T mapFromPages(AbstractPagedInputView source) throws IOException;
```
Comment (JingsongLi) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/typeutils/PagedTypeSerializer.java:82]: Can you remove this method? It is useless but misleading.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/util/collections/binary/BytesHashMapBase.java [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+ * <ul>
+ *   <li>Bytes 0 to 4: len(k)
+ *   <li>Bytes 4 to 4 + len(k): key data
+ *   <li>Bytes 4 + len(k) to 8 + len(k): len(v)
+ *   <li>Bytes 8 + len(k) to 8 + len(k) + len(v): value data
+ * </ul>
+ *
+ * <p>{@code BytesHashMap} are influenced by Apache Spark BytesToBytesMap.
+ */
+public abstract class BytesHashMapBase<K> extends BytesMap<K, BinaryRowData> {
```
Comment (JingsongLi) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/util/collections/binary/BytesHashMapBase.java:69]: NIT: I prefer `AbstractBytesHashMap`

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/util/collections/binary/BytesMap.java [hunk 5]
```diff
@@ -182,9 +176,8 @@ public void reset() {
      *     BinaryRowData form who has only one MemorySegment.
      * @return {@link LookupInfo}
      */
-    public LookupInfo<V> lookup(BinaryRowData key) {
-        // check the looking up key having only one memory segment
-        checkArgument(key.getSegments().length == 1);
+    public LookupInfo<K, V> lookup(K key) {
+        validateKey(key);
```
Comment (JingsongLi) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/util/collections/binary/BytesMap.java:180]: That is legacy code, we don't need to validate it now.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/util/collections/binary/BytesWindowHashMap.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+import org.apache.flink.table.types.logical.LogicalType;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+
+/**
+ * A binary map in the structure like {@code Map<WindowKey, BinaryRowData>}.
+ *
+ * @see BytesHashMapBase for more information about the binary layout.
+ */
+public final class BytesWindowHashMap extends BytesHashMapBase<WindowKey> {
```
Comment (JingsongLi) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/util/collections/binary/BytesWindowHashMap.java:33]: NIT: I prefer `WindowBytesHashMap`, corresponding to `AbstractBytesHashMap`

**Comment by wuchong** [comment]
Thanks for the review @JingsongLi . I have addressed the comments.

**Review by JingsongLi — APPROVED** [review]
Looks good to me.

**Comment by wuchong** [comment]
Thanks @JingsongLi for the reviewing.

**Pull request closed without merging**